### DISCUSSION TITLE
Why can't I get ITM working?

### DISCUSSION THREAD

**Discussion by jeffallen** [oxidecomputer/hubris#334] [Q&A] [answered]
I have been trying to see the output of sys_log, like the ones in jefe (or my own I add), and I can't figure out where I am supposed to see them.

I have noticed the commented out "monitor" lines related to ITM logging in openocd.gdb, and tried experimenting with them. I've gotten the error "SWO pin frequency not set" from OpenOCD. I also discovered "humility ITM", which gives the same error.

I was also kind of expecting to find some kind of "humility log" that streams logs, like "adb logcat" does in the Android world. Any interest in something like that?

Edited to add: I have switched to using semihosting and that worked as expected for the kernel and some messages I added to pong.

**Comment by jeffallen** [comment]
This will likely be fixed with #351. Thanks Bryan.

**Reply by bcantrill** [reply]
Sorry to have missed this! (Somewhat annoyingly, unlike issues and PRs, the GitHub UI doesn't seem to clearly indicate when there are new conversations.)  Depending on your board (do you have a Nucleo H7?), #351 may well be the fix for you; that's out for review, so you should be able to give it a shot by building that branch.  Please let me know if it doesn't work!

**Comment by steveklabnik** [comment]
> I was also kind of expecting to find some kind of "humility log" that streams logs, like "adb logcat" does in the Android world. Any interest in something like that?

I had an old branch working on logging, but it didn't end up landing. At some point, yeah!

**Comment by bcantrill** [comment, marked as answer]
The fix for this (#351) has been integrated. Also, note that [ring buffers](https://github.com/oxidecomputer/hubris/blob/master/lib/ringbuf/src/lib.rs) are -- for a variety of reasons -- preferred to logging for indicating that code was executed.  Ring buffers also enjoy Humility support via [`humliity ringbuf`](https://github.com/oxidecomputer/humility#humility-ringbuf).

**Reply by jeffallen** [reply]
Yes! I came to understand this as I was working on some kind of "string ringbuf" kind of thing. Ringbufs are better because they hold the relevant data, but the formatting code is not compiled into the image stored in flash. It's a bit like the separation of data and presentation, but separated by different domains: embedded domain holds the data in the most compact format possible, and the debugger does the presentation. Classy!
